library/test/src/lib.rs

```diff
@@ -506,7 +506,28 @@ pub fn run_test(
         let supports_threads = !cfg!(target_os = "emscripten") && !cfg!(target_arch = "wasm32");
         if concurrency == Concurrent::Yes && supports_threads {
             let cfg = thread::Builder::new().name(name.as_slice().to_owned());
-            Some(cfg.spawn(runtest).unwrap())
+            let mut runtest = Arc::new(Mutex::new(Some(runtest)));
+            let runtest2 = runtest.clone();
+            cfg.spawn(move || runtest2.lock().unwrap().take().unwrap()())
+                .map_or_else(
+                    |e| {
+                        if e.kind() == io::ErrorKind::WouldBlock {
+                            // `ErrorKind::WouldBlock` means hitting the thread limit on some
+                            // platforms, so run the test synchronously here instead.
+                            Arc::get_mut(&mut runtest) // force rustfmt to break this line
+                                .unwrap()
+                                .get_mut()
+                                .unwrap()
+                                .take()
+                                .unwrap()();
+                            Ok(None)
+                        } else {
+                            Err(e)
+                        }
+                    },
+                    |h| Ok(Some(h)),
+                )
+                .unwrap()
```

Comment:
I admit to struggling a bit with what each of these closures gets and the unwrap here, I think a match would be more readable. Could you do that?